contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java

```diff
@@ -289,14 +237,37 @@ public void build() {
                 conf.setReducerClass(AvroStoreBuilderReducer.class);
             }
 
-            logger.info("Number of chunks: " + numChunks + ", number of reducers: " + numReducers
-                        + ", save keys: " + saveKeys + ", reducerPerBucket: " + reducerPerBucket
-                        + ", buildPrimaryReplicasOnly: " + buildPrimaryReplicasOnly);
             logger.info("Building store...");
-            RunningJob job = JobClient.runJob(conf);
 
-            // Once the job has completed log the counter
-            Counters counters = job.getCounters();
+            // The snipped below copied and adapted from: JobClient.runJob(conf);
+            // We have more control in the error handling this way.
+
+            JobClient jc = new JobClient(conf);
+            RunningJob runningJob = jc.submitJob(conf);
+            Counters counters = runningJob.getCounters();
+
+            try {
+                if (!jc.monitorAndPrintJob(conf, runningJob)) {
+                    // For some datasets, the number of chunks that we calculated is inadequate.
+                    // Here, we try to identify if this is the case.
+                    long mapOutputBytes = counters.getCounter(Task.Counter.MAP_OUTPUT_BYTES);
```

Comment:
You are right that this does not solve the issue of data skew.

In practice, I don't think I've seen a case where data skew was actually the issue that triggered a chunk overflow. It's theoretically possible, but the fact that this category of edge cases can only happen in big data sets (those on the order of: partition count * chunk size) makes it unlikely that there would be significant skew in such big data set.

That being said, I don't think the MR counters can give us per mapper stats out of the box. We could record it ourselves in custom counters, but I don't think it's worth it.

I think we can simply apply a safety margin (like 0.75 or even 0.5) on top of the properly measured chunk size, and we'd be safe enough.